### PR TITLE
[SPARK-26004][SQL] InMemoryTable support StartsWith predicate push down

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

[SPARK-24638](https://issues.apache.org/jira/browse/SPARK-24638) adds support for Parquet file `StartsWith` predicate push down.
`InMemoryTable` can also support this feature.

This is an example to explain how it works, Imagine that the `id` column stored as below:

Partition ID | lowerBound | upperBound
-- | -- | --
p1 | '1' | '9'
p2 | '10' | '19'
p3 | '20' | '29'
p4 | '30' | '39'
p5 | '40' | '49'

A filter ```df.filter($"id".startsWith("2"))``` or ```id like '2%'```
then we substr lowerBound and upperBound:
 
Partition ID | lowerBound.substr(0, Length("2")) | upperBound.substr(0, Length("2"))
-- | -- | --
p1 | '1' | '9'
p2 | '1' | '1'
p3 | '2' | '2'
p4 | '3' | '3'
p5 | '4' | '4'

We can see that we only need to read `p1` and `p3`.

## How was this patch tested?

 unit tests and benchmark tests

benchmark test result:
```
================================================================================================
Pushdown benchmark for StringStartsWith
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
StringStartsWith filter: (value like '10%'): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
InMemoryTable Vectorized                    12068 / 14198          1.3         767.3       1.0X
InMemoryTable Vectorized (Pushdown)           5457 / 8662          2.9         347.0       2.2X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
StringStartsWith filter: (value like '1000%'): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
InMemoryTable Vectorized                      5246 / 5355          3.0         333.5       1.0X
InMemoryTable Vectorized (Pushdown)           2185 / 2346          7.2         138.9       2.4X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_191-b12 on Mac OS X 10.12.6
Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
StringStartsWith filter: (value like '786432%'): Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
InMemoryTable Vectorized                      5112 / 5312          3.1         325.0       1.0X
InMemoryTable Vectorized (Pushdown)           2292 / 2522          6.9         145.7       2.2X
```

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan @HyukjinKwon @kiszk

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 1]
```diff
@@ -237,6 +237,13 @@ case class InMemoryTableScanExec(
       if list.forall(ExtractableLiteral.unapply(_).isDefined) && list.nonEmpty =>
       list.map(l => statsFor(a).lowerBound <= l.asInstanceOf[Literal] &&
         l.asInstanceOf[Literal] <= statsFor(a).upperBound).reduce(_ || _)
+
+    case StartsWith(a: AttributeReference, ExtractableLiteral(l)) =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: can you add some comment to explain it?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 2]
```diff
@@ -237,6 +237,13 @@ case class InMemoryTableScanExec(
...
+      statsFor(a).lowerBound.substr(0, Length(l)) <= l &&
+        l <= statsFor(a).upperBound.substr(0, Length(l))
+    case StartsWith(ExtractableLiteral(l), a: AttributeReference) =>
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: BTW,  a.startswith(b) and b.startswith(a) are not same but why are they same here?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: same question

Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Good question, The last one should be removed, `DataSourceStrategy` has the same logic:  https://github.com/apache/spark/blob/3d6b68b030ee85a0f639dd8e9b68aedf5f27b46f/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L512-L513

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Added to pr description.

**Comment by HyukjinKwon** [comment]
Looks fine to me

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Can you add the comment in the line 240, too?

**Comment by maropu** [comment]
Any update?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: @maropu Done

**Review by maropu — APPROVED** [review]

**Comment by maropu** [comment]
cc: @cloud-fan @HyukjinKwon

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**